File: pkg/earthsci-ast-go/pkg/esm/diagnostic_code_registry_test.go
package esm

// Every diagnostic code this package raises is a registered one.
//
// The code strings are a cross-binding wire contract (esm-spec §9.6.6 calls
// the table "cross-language uniform"), and a code that reaches a raise site
// without a registry constant is exactly how one binding silently drifts from
// the others. This test reads the package's own source with go/parser and
// enforces three things:
//
//  1. A raise site never spells its code as a string literal. A raise site is
//     the first argument of a package constructor whose first parameter is
//     `code string` (newETErr, newClosedFunctionError, ...), a `Code:` /
//     `code:` field in a composite literal, or an assignment to a `.Code` /
//     `.code` field.
//  2. A constant named at a raise site is a registry constant: a package-level
//     string constant whose name begins with Code, code, Error or UnitFinding
//     (codes.go, plus the per-subsystem vocabularies its header lists).
//  3. Every code in the esm-spec §9.6.6 table is registered.
//
// A registry constant that nothing references is reported with t.Logf, not
// failed: CodeSubsystemIndexSetRenameUnsupportedMountForm is kept on purpose so
// the §9.6.6 table stays uniform even though Go never mounts the form that
// raises it.

import (
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"testing"
)

// specDiagnosticCodesExempt lists §9.6.6 codes no binding registers yet.
// `unevaluable_operator` is registered across the bindings by the issue #247
// work (branch claude/issue-247-unevaluable-operator); drop it from this set
// when that lands.
var specDiagnosticCodesExempt = map[string]bool{
	"unevaluable_operator": true,
}

var registryConstName = regexp.MustCompile(`^(Code|code|Error|UnitFinding)[A-Z]`)

type parsedPackage struct {
	fset  *token.FileSet
	files map[string]*ast.File
}

func parseEsmPackageSource(t *testing.T) parsedPackage {
	t.Helper()
	fset := token.NewFileSet()
	entries, err := os.ReadDir(".")
	if err != nil {
		t.Fatal(err)
	}
	files := map[string]*ast.File{}
	for _, e := range entries {
		name := e.Name()
		if e.IsDir() || !strings.HasSuffix(name, ".go") || strings.HasSuffix(name, "_test.go") {
			continue
		}
		f, err := parser.ParseFile(fset, name, nil, 0)
		if err != nil {
			t.Fatalf("parse %s: %v", name, err)
		}
		files[name] = f
	}
	if len(files) == 0 {
		t.Fatal("no package source files found")
	}
	return parsedPackage{fset: fset, files: files}
}

// stringConsts maps every package-level constant with a string-literal value
// to that value.
func stringConsts(pkg parsedPackage) map[string]string {
	out := map[string]string{}
	for _, f := range pkg.files {
		for _, decl := range f.Decls {
			gd, ok := decl.(*ast.GenDecl)
			if !ok || gd.Tok != token.CONST {
				continue
			}
			for _, spec := range gd.Specs {
				vs := spec.(*ast.ValueSpec)
				for i, name := range vs.Names {
					if i >= len(vs.Values) {
						continue
					}
					if lit, ok := vs.Values[i].(*ast.BasicLit); ok && lit.Kind == token.STRING {
						if v, err := strconv.Unquote(lit.Value); err == nil {
							out[name.Name] = v
						}
					}
				}
			}
		}
	}
	return out
}

// codeConstructors returns the package functions whose first parameter is
// `code string`: every argument passed in that slot is a raised code.
func codeConstructors(pkg parsedPackage) map[string]bool {
	out := map[string]bool{}
	for _, f := range pkg.files {
		for _, decl := range f.Decls {
			fd, ok := decl.(*ast.FuncDecl)
			if !ok || fd.Recv != nil || fd.Type.Params == nil || len(fd.Type.Params.List) == 0 {
				continue
			}
			first := fd.Type.Params.List[0]
			typ, ok := first.Type.(*ast.Ident)
			if !ok || typ.Name != "string" || len(first.Names) == 0 || first.Names[0].Name != "code" {
				continue
			}
			out[fd.Name.Name] = true
		}
	}
	return out
}

func isCodeField(name string) bool { return name == "Code" || name == "code" }

// specDiagnosticCodes reads the code column of the esm-spec §9.6.6 table.
func specDiagnosticCodes(t *testing.T) []string {
	t.Helper()
	raw, err := os.ReadFile("../../../../esm-spec.md")
	if err != nil {
		t.Fatalf("read esm-spec.md: %v", err)
	}
	text := string(raw)
	start := strings.Index(text, "\n#### 9.6.6 ")
	if start < 0 {
		t.Fatal("esm-spec.md has no §9.6.6 heading")
	}
	section := text[start+1:]
	if end := strings.Index(section, "\n#### "); end >= 0 {
		section = section[:end]
	}
	row := regexp.MustCompile("(?m)^\\| `([a-z][a-z0-9_]*)` \\|")
	var codes []string
	for _, m := range row.FindAllStringSubmatch(section, -1) {
		codes = append(codes, m[1])
	}
	// Guard the extraction itself: a heading or table-layout change that
	// matched nothing would pass the membership check vacuously.
	if len(codes) < 30 {
		t.Fatalf("extracted only %d codes from the §9.6.6 table; the parser no longer matches the spec layout", len(codes))
	}
	return codes
}

func TestEveryRaisedDiagnosticCodeIsRegistered(t *testing.T) {
	pkg := parseEsmPackageSource(t)
	consts := stringConsts(pkg)
	registry := map[string]string{} // constant name -> code
	for name, value := range consts {
		if registryConstName.MatchString(name) {
			registry[name] = value
		}
	}
	constructors := codeConstructors(pkg)

	var violations []string
	checked := 0
	check := func(expr ast.Expr, context string) {
		pos := pkg.fset.Position(expr.Pos())
		where := pos.Filename + ":" + strconv.Itoa(pos.Line)
		switch e := expr.(type) {
		case *ast.BasicLit:
			if e.Kind != token.STRING {
				return
			}
			checked++
			violations = append(violations, where+": "+context+" raises the literal code "+e.Value+"; name a registry constant (codes.go) instead")
		case *ast.Ident:
			if _, isConst := consts[e.Name]; !isConst {
				return // a variable or parameter passing a code through
			}
			checked++
			if _, ok := registry[e.Name]; !ok {
				violations = append(violations, where+": "+context+" raises "+e.Name+", which is not a registry constant (Code*/Error*/UnitFinding*)")
			}
		}
	}

	for _, name := range registryKeysSorted(pkg.files) {
		ast.Inspect(pkg.files[name], func(n ast.Node) bool {
			switch node := n.(type) {
			case *ast.CallExpr:
				if fn, ok := node.Fun.(*ast.Ident); ok && constructors[fn.Name] && len(node.Args) > 0 {
					check(node.Args[0], fn.Name+"()")
				}
			case *ast.CompositeLit:
				for _, elt := range node.Elts {
					if kv, ok := elt.(*ast.KeyValueExpr); ok {
						if key, ok := kv.Key.(*ast.Ident); ok && isCodeField(key.Name) {
							check(kv.Value, key.Name+": field")
						}
					}
				}
			case *ast.AssignStmt:
				for i, lhs := range node.Lhs {
					if sel, ok := lhs.(*ast.SelectorExpr); ok && isCodeField(sel.Sel.Name) && i < len(node.Rhs) {
						check(node.Rhs[i], "."+sel.Sel.Name+" assignment")
					}
				}
			}
			return true
		})
	}

	// Guard the scan: a layout drift that matched no raise site would pass
	// vacuously.
	if checked < 100 {
		t.Fatalf("scanned only %d raise sites; the raise-site patterns no longer match the source", checked)
	}
	sort.Strings(violations)
	for _, v := range violations {
		t.Error(v)
	}

	registered := map[string]bool{}
	for _, value := range registry {
		registered[value] = true
	}
	for _, code := range specDiagnosticCodes(t) {
		if !registered[code] && !specDiagnosticCodesExempt[code] {
			t.Errorf("esm-spec §9.6.6 code %q has no registry constant in this binding", code)
		}
	}

	// Registered but never referenced: advisory only. The declaration itself
	// is one occurrence of the name, so a referenced constant occurs twice.
	occurrences := map[string]int{}
	for _, f := range pkg.files {
		ast.Inspect(f, func(n ast.Node) bool {
			if id, ok := n.(*ast.Ident); ok {
				occurrences[id.Name]++
			}
			return true
		})
	}
	for _, name := range registryKeysSorted(registry) {
		if occurrences[name] < 2 {
			t.Logf("warning: registry constant %s (%q) is never raised", name, registry[name])
		}
	}
}

func registryKeysSorted[V any](m map[string]V) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

File: pkg/earthsci-ast-ts/src/diagnostic-code-registry.test.ts
/**
 * Every diagnostic code this binding raises is a registered one.
 *
 * The code strings are a cross-binding wire contract (esm-spec §9.6.6 calls
 * the table "cross-language uniform"), and a code that reaches a raise site
 * without going through `ERROR_CODES` is exactly how one binding silently
 * drifts from the others. This test parses `src/` with the TypeScript compiler
 * API and enforces three things:
 *
 *  1. A raise site never spells its code as a string literal. A raise site is
 *     the argument bound to a parameter named `code` (of a constructor —
 *     including through `super(...)` — a function, or a method), that
 *     parameter's default value, a `code:` property in an object literal, a
 *     class property named `code`, or an assignment to `.code`.
 *  2. An `ERROR_CODES.<KEY>` reference at a raise site names a key the registry
 *     has.
 *  3. Every code in the esm-spec §9.6.6 table is in `ERROR_CODES`.
 *
 * A registry entry nothing references is reported with `console.warn`, not
 * failed.
 */

import * as fs from 'node:fs'
import * as path from 'node:path'
import { fileURLToPath } from 'node:url'
import ts from 'typescript'
import { describe, expect, it } from 'vitest'
import { ERROR_CODES } from './errors.js'

const SRC = path.dirname(fileURLToPath(import.meta.url))
const SPEC = path.resolve(SRC, '../../../esm-spec.md')

function sourceFiles(dir: string): string[] {
  const out: string[] = []
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name)
    if (entry.isDirectory()) out.push(...sourceFiles(full))
    else if (full.endsWith('.ts') && !full.endsWith('.test.ts') && !full.endsWith('.d.ts'))
      out.push(full)
  }
  return out.sort()
}

function specDiagnosticCodes(): string[] {
  const text = fs.readFileSync(SPEC, 'utf8')
  const start = text.indexOf('\n#### 9.6.6 ')
  expect(start, 'esm-spec.md has no §9.6.6 heading').toBeGreaterThanOrEqual(0)
  let section = text.slice(start + 1)
  const end = section.indexOf('\n#### ')
  if (end >= 0) section = section.slice(0, end)
  return [...section.matchAll(/^\| `([a-z][a-z0-9_]*)` \|/gm)].map((m) => m[1]!)
}

interface Scan {
  violations: string[]
  checked: number
  referencedKeys: Set<string>
}

function scan(): Scan {
  const parsed = sourceFiles(SRC).map((file) =>
    ts.createSourceFile(file, fs.readFileSync(file, 'utf8'), ts.ScriptTarget.Latest, true),
  )

  // Pass 1: which parameter position carries the code, per callable name, and
  // each class's superclass (so a subclass without its own constructor, and a
  // `super(...)` call, resolve to the constructor that declares `code`).
  const ctorCodeIndex = new Map<string, number>()
  // Functions resolve within their own file: two files may each declare a local
  // helper of the same name, and only one of them may take a `code`.
  const fnCodeIndex = new Map<string, number>() // `${file}\0${name}`
  const superclass = new Map<string, string>()
  const codeParamIndex = (params: ts.NodeArray<ts.ParameterDeclaration>) =>
    params.findIndex((p) => ts.isIdentifier(p.name) && p.name.text === 'code')
  for (const sf of parsed) {
    const visit = (node: ts.Node): void => {
      if (ts.isClassDeclaration(node) && node.name) {
        const ext = node.heritageClauses?.find((h) => h.token === ts.SyntaxKind.ExtendsKeyword)
        const base = ext?.types[0]?.expression
        if (base && ts.isIdentifier(base)) superclass.set(node.name.text, base.text)
      }
      if (
        ts.isConstructorDeclaration(node) &&
        ts.isClassDeclaration(node.parent) &&
        node.parent.name
      ) {
        const i = codeParamIndex(node.parameters)
        if (i >= 0) ctorCodeIndex.set(node.parent.name.text, i)
      }
      if (
        (ts.isFunctionDeclaration(node) || ts.isMethodDeclaration(node)) &&
        node.name &&
        ts.isIdentifier(node.name)
      ) {
        const i = codeParamIndex(node.parameters)
        if (i >= 0) fnCodeIndex.set(`${sf.fileName}\0${node.name.text}`, i)
      }
      // A local helper bound to a name (`const warn = (message, code) => ...`)
      // is a raise site exactly like a declared function.
      if (
        ts.isVariableDeclaration(node) &&
        ts.isIdentifier(node.name) &&
        node.initializer &&
        (ts.isArrowFunction(node.initializer) || ts.isFunctionExpression(node.initializer))
      ) {
        const i = codeParamIndex(node.initializer.parameters)
        if (i >= 0) fnCodeIndex.set(`${sf.fileName}\0${node.name.text}`, i)
      }
      ts.forEachChild(node, visit)
    }
    visit(sf)
  }
  const classesWithCtor = new Set<string>()
  for (const sf of parsed) {
    const visit = (node: ts.Node): void => {
      if (
        ts.isConstructorDeclaration(node) &&
        ts.isClassDeclaration(node.parent) &&
        node.parent.name
      ) {
        classesWithCtor.add(node.parent.name.text)
      }
      ts.forEachChild(node, visit)
    }
    visit(sf)
  }
  const resolveCtor = (cls: string | undefined): number | undefined => {
    for (let c = cls, guard = 0; c && guard < 32; c = superclass.get(c), guard++) {
      if (classesWithCtor.has(c)) return ctorCodeIndex.get(c)
    }
    return undefined
  }

  const result: Scan = { violations: [], checked: 0, referencedKeys: new Set() }

  // Pass 2: check every raise site.
  for (const sf of parsed) {
    const rel = path.relative(SRC, sf.fileName)
    const fileStringConsts = new Map<string, string>()
    for (const stmt of sf.statements) {
      if (!ts.isVariableStatement(stmt)) continue
      for (const decl of stmt.declarationList.declarations) {
        if (
          ts.isIdentifier(decl.name) &&
          decl.initializer &&
          ts.isStringLiteralLike(decl.initializer)
        ) {
          fileStringConsts.set(decl.name.text, decl.initializer.text)
        }
      }
    }

    const check = (expr: ts.Expression | undefined, context: string): void => {
      if (!expr) return
      while (
        ts.isParenthesizedExpression(expr) ||
        ts.isAsExpression(expr) ||
        ts.isSatisfiesExpression(expr)
      ) {
        expr = expr.expression
      }
      const literal = ts.isStringLiteralLike(expr)
        ? expr.text
        : ts.isIdentifier(expr)
          ? fileStringConsts.get(expr.text)
          : undefined
      // The uppercase `E_*` names (E_REF_*, E_CANONICAL_*) are a separate
      // stable error-NAME vocabulary, excluded from the snake_case code
      // registry by the same convention Julia (`E_TREEWALK_*`) and Rust (the DAE
      // lowering's `E_*`) follow; they are not diagnostic codes.
      if (literal !== undefined && /^E_[A-Z0-9_]+$/.test(literal)) return
      const line = sf.getLineAndCharacterOfPosition(expr.getStart(sf)).line + 1
      const where = `${rel}:${line}: ${context}`
      if (ts.isStringLiteralLike(expr)) {
        result.checked++
        result.violations.push(
          `${where} raises the literal code '${expr.text}'; use ERROR_CODES instead`,
        )
      } else if (
        ts.isPropertyAccessExpression(expr) &&
        ts.isIdentifier(expr.expression) &&
        expr.expression.text === 'ERROR_CODES'
      ) {
        result.checked++
        if (!Object.prototype.hasOwnProperty.call(ERROR_CODES, expr.name.text)) {
          result.violations.push(
            `${where} raises ERROR_CODES.${expr.name.text}, which the registry does not define`,
          )
        }
      } else if (ts.isIdentifier(expr) && fileStringConsts.has(expr.text)) {
        result.checked++
        result.violations.push(
          `${where} raises the module constant ${expr.text} = '${fileStringConsts.get(expr.text)}'; use ERROR_CODES instead`,
        )
      }
    }

    const enclosingClass = (node: ts.Node): ts.ClassLikeDeclaration | undefined => {
      for (let n = node.parent; n; n = n.parent) if (ts.isClassLike(n)) return n
      return undefined
    }

    const visit = (node: ts.Node): void => {
      if (
        ts.isPropertyAccessExpression(node) &&
        ts.isIdentifier(node.expression) &&
        node.expression.text === 'ERROR_CODES'
      ) {
        result.referencedKeys.add(node.name.text)
      }
      if (ts.isNewExpression(node) && ts.isIdentifier(node.expression)) {
        const i = resolveCtor(node.expression.text)
        if (i !== undefined) check(node.arguments?.[i], `new ${node.expression.text}()`)
      } else if (ts.isCallExpression(node)) {
        if (node.expression.kind === ts.SyntaxKind.SuperKeyword) {
          const cls = enclosingClass(node)
          const i = resolveCtor(cls?.name ? superclass.get(cls.name.text) : undefined)
          if (i !== undefined) check(node.arguments[i], 'super()')
        } else {
          const callee = ts.isIdentifier(node.expression)
            ? node.expression.text
            : ts.isPropertyAccessExpression(node.expression)
              ? node.expression.name.text
              : undefined
          const i = callee === undefined ? undefined : fnCodeIndex.get(`${sf.fileName}\0${callee}`)
          if (i !== undefined) check(node.arguments[i], `${callee}()`)
        }
      } else if (ts.isParameter(node) && ts.isIdentifier(node.name) && node.name.text === 'code') {
        check(node.initializer, 'default value of a `code` parameter')
      } else if (ts.isPropertyAssignment(node) && node.name.getText(sf) === 'code') {
        check(node.initializer, '`code:` property')
      } else if (ts.isPropertyDeclaration(node) && node.name.getText(sf) === 'code') {
        check(node.initializer, 'class property `code`')
      } else if (
        ts.isBinaryExpression(node) &&
        node.operatorToken.kind === ts.SyntaxKind.EqualsToken &&
        ts.isPropertyAccessExpression(node.left) &&
        node.left.name.text === 'code'
      ) {
        check(node.right, '`.code` assignment')
      }
      ts.forEachChild(node, visit)
    }
    visit(sf)
  }
  return result
}

describe('diagnostic code registry coverage', () => {
  const result = scan()

  it('every raise site names its code through ERROR_CODES', () => {
    // Guard the scan: a layout drift that matched no raise site would pass
    // vacuously.
    expect(result.checked).toBeGreaterThanOrEqual(100)
    expect(result.violations).toEqual([])
  })

  it('registers every code in the esm-spec §9.6.6 table', () => {
    const codes = specDiagnosticCodes()
    // Guard the extraction: a heading or table-layout change that matched
    // nothing would pass the membership check vacuously.
    expect(codes.length).toBeGreaterThanOrEqual(30)
    const registered = new Set<string>(Object.values(ERROR_CODES))
    expect(codes.filter((c) => !registered.has(c))).toEqual([])
  })

  it('reports registry entries nothing raises (advisory)', () => {
    const unused = Object.keys(ERROR_CODES).filter((k) => !result.referencedKeys.has(k))
    if (unused.length > 0) {
      console.warn(`ERROR_CODES entries never referenced from src/: ${unused.join(', ')}`)
    }
  })
})
